erizo/src/erizo/MediaStream.cpp

```diff
@@ -0,0 +1,515 @@
+
+  audioEnabled_ = remote_sdp_->hasAudio;
+  videoEnabled_ = remote_sdp_->hasVideo;
+
+  rtcp_processor_->addSourceSsrc(getAudioSourceSSRC());
+  std::for_each(video_source_ssrc_list_.begin(), video_source_ssrc_list_.end(), [this] (uint32_t new_ssrc){
+      rtcp_processor_->addSourceSsrc(new_ssrc);
+  });
+
+  initializePipeline();
+
+  return true;
+}
+
+bool MediaStream::setLocalSdp(std::shared_ptr<SdpInfo> sdp) {
+  local_sdp_ = sdp;
+  return true;
+}
+
+void MediaStream::initializePipeline() {
+  pipeline_->addService(shared_from_this());
+  pipeline_->addService(rtcp_processor_);
+  pipeline_->addService(stats_);
+  pipeline_->addService(quality_manager_);
+  pipeline_->addService(packet_buffer_);
+
+  pipeline_->addFront(PacketReader(this));
+
+  pipeline_->addFront(LayerDetectorHandler());
+  pipeline_->addFront(RtcpProcessorHandler());
+  pipeline_->addFront(FecReceiverHandler());
+  pipeline_->addFront(LayerBitrateCalculationHandler());
+  pipeline_->addFront(QualityFilterHandler());
+  pipeline_->addFront(IncomingStatsHandler());
+  pipeline_->addFront(RtpTrackMuteHandler());
+  pipeline_->addFront(RtpSlideShowHandler());
+  pipeline_->addFront(RtpPaddingGeneratorHandler());
+  pipeline_->addFront(PliPacerHandler());
+  pipeline_->addFront(BandwidthEstimationHandler());
+  pipeline_->addFront(RtpPaddingRemovalHandler());
+  pipeline_->addFront(RtcpFeedbackGenerationHandler());
+  pipeline_->addFront(RtpRetransmissionHandler());
+  pipeline_->addFront(SRPacketHandler());
+  pipeline_->addFront(SenderBandwidthEstimationHandler());
+  pipeline_->addFront(OutgoingStatsHandler());
+
+  pipeline_->addFront(PacketWriter(this));
+  pipeline_->finalize();
+  pipeline_initialized_ = true;
+}
+
+int MediaStream::deliverAudioData_(std::shared_ptr<DataPacket> audio_packet) {
+  if (audioEnabled_ == true) {
+    sendPacketAsync(std::make_shared<DataPacket>(*audio_packet));
+  }
+  return audio_packet->length;
+}
+
+int MediaStream::deliverVideoData_(std::shared_ptr<DataPacket> video_packet) {
+  if (videoEnabled_ == true) {
+    sendPacketAsync(std::make_shared<DataPacket>(*video_packet));
+  }
+  return video_packet->length;
+}
+
+int MediaStream::deliverFeedback_(std::shared_ptr<DataPacket> fb_packet) {
+  RtcpHeader *chead = reinterpret_cast<RtcpHeader*>(fb_packet->data);
+  uint32_t recvSSRC = chead->getSourceSSRC();
+  if (isVideoSourceSSRC(recvSSRC)) {
+    fb_packet->type = VIDEO_PACKET;
+    sendPacketAsync(fb_packet);
+  } else if (isAudioSourceSSRC(recvSSRC)) {
+    fb_packet->type = AUDIO_PACKET;
+    sendPacketAsync(fb_packet);
+  } else {
+    ELOG_DEBUG("%s deliverFeedback unknownSSRC: %u, localVideoSSRC: %u, localAudioSSRC: %u",
+                toLog(), recvSSRC, this->getVideoSourceSSRC(), this->getAudioSourceSSRC());
+  }
+  return fb_packet->length;
+}
+
+int MediaStream::deliverEvent_(MediaEventPtr event) {
+  auto conn_ptr = shared_from_this();
+  worker_->task([conn_ptr, event]{
```

Comment:
I now think this async task is not completely safe, we might need to switch to weak_ptrs here, but in another PR